### PR TITLE
Implement InlineFunction to force inline at call site

### PR DESCRIPTION
**Pull request by hardikar**
A simple wrapper around llvm::InlineFunction, with some unit tests.
@karthijrk @armenatzoglou @gcaragea @foyzur Please take a look.

### CODE REVIEW

File: src/backend/codegen/include/codegen/utils/codegen_utils.h [hunk 1]
```diff
@@ -411,6 +413,18 @@ class CodegenUtils {
...
 
+  /*
+   * @brief Force inline an LLVM function at the call site. Note that this only
+   *        does one level of inlining.
+   *
+   * @param LLVM call Instruction where the called function should be inlined
+   * @return false if it is not possible to inline. true otherwise.
+   */
+  bool InlineFunction(llvm::CallInst* call_inst) {
+    llvm::InlineFunctionInfo info;
```
Comment (karthijrk) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: This variable is just declared and passed to llvm function. Should we need to initialize something. I am just wondering why we need to pass if we don't do anything to that object.

File: src/backend/codegen/tests/codegen_utils_unittest.cc [hunk 2]
```diff
@@ -2758,6 +2759,60 @@ TEST_F(CodegenUtilsTest, GetOrGetOrRegisterExternalFunctionTest) {
...
+  irb->CreateRet(second_call);
+
+
+  EXPECT_EQ(GetLLVMFunctionCallCount(add_two_fn), 2);
+
+  EXPECT_TRUE(codegen_utils_->InlineFunction(first_call));
+  EXPECT_EQ(GetLLVMFunctionCallCount(add_two_fn), 1);
+
+#ifdef GPCODEGEN_DEBUG
+  EXPECT_DEATH(codegen_utils_->InlineFunction(first_call));
```
Comment (karthijrk) [inline src/backend/codegen/tests/codegen_utils_unittest.cc]: Will this work if underlying llvm is built in release mode with no assert ?

Comment (hardikar) [inline src/backend/codegen/tests/codegen_utils_unittest.cc]: Then we don't do GPCODEGEN_DEBUG right? I thought that's why we introduced that #define ?

Comment (karthijrk) [inline src/backend/codegen/tests/codegen_utils_unittest.cc]: It depends on the llvm library we link. I am not sure in make file we detect that and set appropriately

File: src/backend/codegen/include/codegen/utils/codegen_utils.h (hunk 1 above)
Comment (hardikar) [inline src/backend/codegen/include/codegen/utils/codegen_utils.h]: We use the default constructor of the InlineFunctionInfo. It is passed to InlineFunction as a reference, and it's [fields](http://llvm.org/docs/doxygen/html/Cloning_8h_source.html#l00179) are updated but InlineFunctions. But we mostly don't care about it.

File: src/backend/codegen/tests/codegen_utils_unittest.cc (hunk 2 above)
Comment (hardikar) [inline src/backend/codegen/tests/codegen_utils_unittest.cc]: I see to answer your original question - the EXPECT_DEATH comes from and assert somewhere in LLVM. But it looks like we segfault there instead. And I'm not sure yet why.

Comment (hardikar) [inline src/backend/codegen/tests/codegen_utils_unittest.cc]: This is the assertion : Assertion failed: (TheCall->getParent() && TheCall->getParent()->getParent() && "Instruction not in function!"), function InlineFunction, file /Users/shardikar/workplace/orca/llvm-3.7.1.src/lib/Transforms/Utils/InlineFunction.cpp, line 923.

Comment (hardikar) [inline src/backend/codegen/tests/codegen_utils_unittest.cc]: @karthijrk I changed the InlineFunction to catch the case when we call it with a CallInst that doesn't exist. The LLVM doesn't do this check on non-assert compiled builds, so we can do it ourselves. Have a look again.

**Comment by karthijrk** [comment]
LGTM 👍

**Pull request merged**